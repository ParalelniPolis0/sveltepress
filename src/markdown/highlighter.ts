import { readFileSync } from 'fs'
import { resolve } from 'path'
import { fileURLToPath } from 'url'
import { getHighlighter } from 'shiki'
import type { Highlighter } from '@svelte-press/vite'

const __dirname = fileURLToPath(new URL('.', import.meta.url))

export const COMMAND_RE = /\/\/ \[svp\! (hl(:-?\d+(,-?\d+)?)?)\]/

const nightOwl = JSON.parse(readFileSync(resolve(__dirname, './night-owl.json'), 'utf-8'))
const vitesseLight = JSON.parse(readFileSync(resolve(__dirname, './vitesse-light.json'), 'utf-8'))

const createHighlightWith: (theme: string) => Highlighter = theme => (code, lang) => getHighlighter({
  theme,
  langs: ['svelte', 'sh', 'js', 'html', 'ts'],
}).then(
  shikiHighlighter => shikiHighlighter
    .codeToHtml(code, { lang })
    .replace(/\{/g, '&#123;')
    .replace(/\}/g, '&#125;'),
)

const highlighterDark = createHighlightWith(nightOwl)

const highlighterLight = createHighlightWith(vitesseLight)

export function getCommand(line: string) {
  const matches = COMMAND_RE.exec(line)
  if (matches && matches.length) {
    const [comment] = matches
    return [comment.replace(/^\/\/ \[svp\! /, '').replace(/\]$/, ''), line.replace(COMMAND_RE, '')]
  }

  return ['', line]
}

export function highlightLine(startEnd: string, idx: number) {
  if (!startEnd)
    return `<div class="absolute left-0 right-0 z-2 h-[1.5em] bg-[rgba(0,0,0,0.1)] dark:bg-[rgba(255,255,255,0.1)] svp-code-highlight-line" style="top: calc(${idx * 1.5}em + 12px);"></div>`
}

const highlighter: Highlighter = async (code, lang) => {
  const commandDoms = []
  if (lang !== 'md') {
    code = code.split('\n').map((line, i) => {
      const [command, newLine] = getCommand(line)
      if (!command)
        return line

      const [name, params] = command.split(':')
      switch (name) {
        case 'hl':
          commandDoms.push(highlightLine(params, i))
      }
      return newLine
    }).join('\n')
  }
  return `<div class="relative bg-white dark:bg-[#011627] p-[12px] rounded text-[14px]">
      ${commandDoms.join('\n')}
      ${await highlighterLight(code, lang)}
      ${await highlighterDark(code, lang)}
      <div class="absolute top-2 right-3 text-cool-gray-3 dark:text-cool-gray-7 text-[12px] z-2">
        ${lang}
      </div>
  </div>`
}

export default highlighter
