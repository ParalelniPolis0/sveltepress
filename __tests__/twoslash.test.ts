import { describe, expect, it } from 'vitest'
import highlighter from '../src/markdown/highlighter'

describe('twoslash', () => {
  it('renderer floating svelte', async () => {
    const code = `
const num = 1

const foo = 'bar'

`

    expect(await highlighter(code, 'ts')).toMatchSnapshot()
  })
})
