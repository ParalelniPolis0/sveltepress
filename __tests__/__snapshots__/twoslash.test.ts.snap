// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`twoslash > renderer floating svelte 1`] = `
"
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl twoslash lsp" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> </span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"><Floating class="twoslash-hover"><span>num</span><svelte:fragment slot="floating-content"><span class="twoslash-popup-container"><code class="twoslash-popup-code"><span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> num</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#2F798A;--shiki-dark:#F78C6C"> 1</span></span></code></span></svelte:fragment></Floating></span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#2F798A;--shiki-dark:#F78C6C"> 1</span></span>
<span class="line"></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> </span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"><Floating class="twoslash-hover"><span>foo</span><svelte:fragment slot="floating-content"><span class="twoslash-popup-container"><code class="twoslash-popup-code"><span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> foo</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#B5695999;--shiki-dark:#D9F5DD"> "</span><span style="color:#B56959;--shiki-dark:#ECC48D">bar</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span></span></code></span></svelte:fragment></Floating></span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#B5695999;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">bar</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">'</span></span>
<span class="line"></span>
<span class="line"></span></code></pre>
    <div class="svp-code-block--lang">
      ts
    </div>
    <CopyCode />
    
  </div>
</div>"
`;

exports[`twoslash > twoslash svelte 1`] = `
"
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl twoslash lsp" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#B07D48;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> context</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#B56959;--shiki-dark:#ECC48D">module</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#B07D48;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> lang</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#B56959;--shiki-dark:#ECC48D">ts</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"><span style="color:#1E754F;--shiki-dark:#C792EA;font-style:inherit;--shiki-dark-font-style:italic">  export</span><span style="color:#AB5959;--shiki-dark:#C792EA"> const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> </span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"><Floating class="twoslash-hover"><span>FOO</span><svelte:fragment slot="floating-content"><span class="twoslash-popup-container"><code class="twoslash-popup-code"><span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">const</span><span style="color:#B07D48;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> FOO</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#B5695999;--shiki-dark:#D9F5DD"> "</span><span style="color:#B56959;--shiki-dark:#ECC48D">BAR</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span></span></code></span></svelte:fragment></Floating></span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#B5695999;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">BAR</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">'</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;/</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#B07D48;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> lang</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#B56959;--shiki-dark:#ECC48D">ts</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">"</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"><span style="color:#1E754F;--shiki-dark:#C792EA;font-style:inherit;--shiki-dark-font-style:italic">  import</span><span style="color:#999999;--shiki-dark:#D6DEEB"> &#123;</span><span style="color:#B07D48;--shiki-dark:#D6DEEB"> </span><span style="color:#B07D48;--shiki-dark:#D6DEEB"><Floating class="twoslash-hover"><span>onMount</span><svelte:fragment slot="floating-content"><span class="twoslash-popup-container"><code class="twoslash-popup-code"><span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">function</span><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"> onMount</span><span style="color:#999999;--shiki-dark:#D6DEEB">&#x3C;</span><span style="color:#2E8F82;--shiki-dark:#FFCB8B">T</span><span style="color:#999999;--shiki-dark:#D6DEEB">></span><span style="color:#999999;--shiki-dark:#D9F5DD">(</span><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic">fn</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#999999;--shiki-dark:#D9F5DD"> ()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#2E8F82;--shiki-dark:#FFCB8B"> NotFunction</span><span style="color:#999999;--shiki-dark:#D6DEEB">&#x3C;</span><span style="color:#2E8F82;--shiki-dark:#FFCB8B">T</span><span style="color:#999999;--shiki-dark:#D6DEEB">> </span><span style="color:#999999;--shiki-dark:#7FDBCA">|</span><span style="color:#2E8F82;--shiki-dark:#FFCB8B"> Promise</span><span style="color:#999999;--shiki-dark:#D6DEEB">&#x3C;</span><span style="color:#2E8F82;--shiki-dark:#FFCB8B">NotFunction</span><span style="color:#999999;--shiki-dark:#D6DEEB">&#x3C;</span><span style="color:#2E8F82;--shiki-dark:#FFCB8B">T</span><span style="color:#999999;--shiki-dark:#D6DEEB">>> </span><span style="color:#999999;--shiki-dark:#7FDBCA">|</span><span style="color:#999999;--shiki-dark:#D6DEEB"> (</span><span style="color:#999999;--shiki-dark:#D9F5DD">()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#2E8F82;--shiki-dark:#C5E478"> any</span><span style="color:#999999;--shiki-dark:#D6DEEB">)</span><span style="color:#999999;--shiki-dark:#D9F5DD">)</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#2E8F82;--shiki-dark:#C5E478"> void</span></span></code><div class="twoslash-popup-docs"><p>The <code>onMount</code> function schedules a callback to run as soon as the component has been mounted to the DOM.
It must be called during the component's initialisation (but doesn't need to live <em>inside</em> the component;
it can be called from an external module).</p>
<p>If a function is returned <em>synchronously</em> from <code>onMount</code>, it will be called when the component is unmounted.</p>
<p><code>onMount</code> does not run inside a <a href="https://svelte.dev/docs#run-time-server-side-component-api">server-side component</a>.</p>
<p><a href="https://svelte.dev/docs/svelte#onmount">https://svelte.dev/docs/svelte#onmount</a></p></div></span></svelte:fragment></Floating></span><span style="color:#999999;--shiki-dark:#D6DEEB"> &#125;</span><span style="color:#1E754F;--shiki-dark:#C792EA;font-style:inherit;--shiki-dark-font-style:italic"> from</span><span style="color:#B5695999;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">svelte</span><span style="color:#B5695999;--shiki-dark:#D9F5DD">'</span></span>
<span class="line"></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">  let</span><span style="color:#B07D48;--shiki-dark:#D6DEEB;font-style:inherit;--shiki-dark-font-style:italic"> </span><span style="color:#B07D48;--shiki-dark:#D6DEEB;font-style:inherit;--shiki-dark-font-style:italic"><Floating class="twoslash-hover"><span>count</span><svelte:fragment slot="floating-content"><span class="twoslash-popup-container"><code class="twoslash-popup-code"><span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">let</span><span style="color:#B07D48;--shiki-dark:#D6DEEB;font-style:inherit;--shiki-dark-font-style:italic"> count</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#2E8F82;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> number</span></span></code></span></svelte:fragment></Floating></span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#2F798A;--shiki-dark:#F78C6C"> 0</span></span>
<span class="line"></span>
<span class="line"><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic">  </span><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic"><Floating class="twoslash-hover"><span>onMount</span><svelte:fragment slot="floating-content"><span class="twoslash-popup-container"><code class="twoslash-popup-code"><span class="line"><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic">onMount</span><span style="color:#999999;--shiki-dark:#D6DEEB">&#x3C;</span><span style="color:#2E8F82;--shiki-dark:#C5E478">void</span><span style="color:#999999;--shiki-dark:#D6DEEB">>(</span><span style="color:#B07D48;--shiki-dark:#D6DEEB">fn</span><span style="color:#393A34;--shiki-dark:#D6DEEB">: </span><span style="color:#999999;--shiki-dark:#D9F5DD">()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#AB5959;--shiki-dark:#7FDBCA"> void</span><span style="color:#AB5959;--shiki-dark:#C792EA"> |</span><span style="color:#999999;--shiki-dark:#D6DEEB"> (</span><span style="color:#999999;--shiki-dark:#D9F5DD">()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#B07D48;--shiki-dark:#D6DEEB"> any</span><span style="color:#999999;--shiki-dark:#D6DEEB">)</span><span style="color:#AB5959;--shiki-dark:#C792EA"> |</span><span style="color:#998418;--shiki-dark:#C5E478"> Promise</span><span style="color:#999999;--shiki-dark:#C792EA">&#x3C;</span><span style="color:#AB5959;--shiki-dark:#7FDBCA">void</span><span style="color:#999999;--shiki-dark:#C792EA">></span><span style="color:#999999;--shiki-dark:#D6DEEB">)</span><span style="color:#393A34;--shiki-dark:#D6DEEB">: </span><span style="color:#AB5959;--shiki-dark:#7FDBCA">void</span></span></code><div class="twoslash-popup-docs"><p>The <code>onMount</code> function schedules a callback to run as soon as the component has been mounted to the DOM.
It must be called during the component's initialisation (but doesn't need to live <em>inside</em> the component;
it can be called from an external module).</p>
<p>If a function is returned <em>synchronously</em> from <code>onMount</code>, it will be called when the component is unmounted.</p>
<p><code>onMount</code> does not run inside a <a href="https://svelte.dev/docs#run-time-server-side-component-api">server-side component</a>.</p>
<p><a href="https://svelte.dev/docs/svelte#onmount">https://svelte.dev/docs/svelte#onmount</a></p></div></span></svelte:fragment></Floating></span><span style="color:#999999;--shiki-dark:#D6DEEB">(</span><span style="color:#999999;--shiki-dark:#D9F5DD">()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#999999;--shiki-dark:#D6DEEB"> &#123;</span></span>
<span class="line"><span style="color:#A0ADA0;--shiki-dark:#637777">    //</span><span style="color:#A0ADA0;--shiki-dark:#637777;font-style:inherit;--shiki-dark-font-style:italic"> mounted</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#D6DEEB">  &#125;)</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;/</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;</span><span style="color:#1E754F;--shiki-dark:#CAECE6">button</span><span style="color:#1E754F;--shiki-dark:#C792EA;font-style:inherit;--shiki-dark-font-style:italic"> on</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#2E8F82;--shiki-dark:#7FDBCA">cl</span><span style="color:#2E8F82;--shiki-dark:#7FDBCA">ick</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#999999;--shiki-dark:#D3423E">&#123;</span><span style="color:#999999;--shiki-dark:#D9F5DD">()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#B07D48;--shiki-dark:#D6DEEB"> </span><span style="color:#B07D48;--shiki-dark:#D6DEEB"><Floating class="twoslash-hover"><span>count</span><svelte:fragment slot="floating-content"><span class="twoslash-popup-container"><code class="twoslash-popup-code"><span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">let</span><span style="color:#B07D48;--shiki-dark:#D6DEEB;font-style:inherit;--shiki-dark-font-style:italic"> count</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#2E8F82;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> number</span></span></code></span></svelte:fragment></Floating></span><span style="color:#AB5959;--shiki-dark:#C792EA">++</span><span style="color:#999999;--shiki-dark:#D3423E">&#125;</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"><span style="color:#393A34;--shiki-dark:#D6DEEB">  Count is: </span><span style="color:#999999;--shiki-dark:#D3423E">&#123;</span><span style="color:#B07D48;--shiki-dark:#D6DEEB"><Floating class="twoslash-hover"><span>count</span><svelte:fragment slot="floating-content"><span class="twoslash-popup-container"><code class="twoslash-popup-code"><span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">let</span><span style="color:#B07D48;--shiki-dark:#D6DEEB;font-style:inherit;--shiki-dark-font-style:italic"> count</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#2E8F82;--shiki-dark:#C5E478;font-style:inherit;--shiki-dark-font-style:italic"> number</span></span></code></span></svelte:fragment></Floating></span><span style="color:#999999;--shiki-dark:#D3423E">&#125;</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;/</span><span style="color:#1E754F;--shiki-dark:#CAECE6">button</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"></span></code></pre>
    <div class="svp-code-block--lang">
      svelte
    </div>
    <CopyCode />
    
  </div>
</div>"
`;
